### DISCUSSION TITLE
A more dense way to handle exceptions

### DISCUSSION THREAD

**Discussion by sascha-andres** [dotnet/csharplang#7131] [Language Ideas]
# Abstract

To improve readability common approaches are to limit nesting and to reduce line of code required to code a feature. Error handling usually adds nesting and lines of code. Using exceptions as a way to handle errors is the de facto standard for a wide range of languages including C# itself. I'm proposing a new syntax that can be lowered to the current syntax but will make the code more readable.

# Proposal

We propose adding a new `catch with` statement to the language. We propose the syntax

    catch with <delegate>

This would register <delegate> as an exception handler from the current line until the end of the enclosing block. <delegate> must
be a delegate of type Func(E) where E is a subtype of Exception.

The compiler would lower it to a try/catch statement, by enclosing all affected lines in a catch block and moving <delegate> to the catch block.

For example

    void Method() {
      catch with => (e) {
        // do something with the exception
      }
      SomeMethodThrowingAnException();
      Console.WriteLine("done")
    }

would be lowered to:

    void Method() {
      try {
        SomeMethodThrowingAnException();
        Console.WriteLine("done");
      } catch (Exception ex) {
        // do something with the exception
      }
    }

As the actual code implementing the error handling is expressed we could also assign a method from the current instance or a static method:

    void ErrorHandler(Exception ex) {
      // do something with the exception
    }

    void Method() {
      catch with => ErrorHandler;
      SomeMethodThrowingAnException();
      Console.WriteLine("done")
    }

Which could be lowered either to call the method inside the catch or by inlining the code from the method. The handler could also just re-throw the exception.

Error handler could be nested for different ways to handle errors according to local requirements:

    void Method() {
      catch with => ErrorHandler;
      SomeMethodThrowingAnException();
      catch with => AnotherErrorHandler;
      SomeMethodThrowingAnException();
    }

This would be lowered to the following code:

    void Method() {
      try {
        SomeMethodThrowingAnException();
        try {
          SomeMethodThrowingAnException();
        } catch (Exception ex2) {
          // do something else
        }
      } catch (Exception ex) {
        // do something
      }
    }

Handling should be limited to the current scope. When defined on a method level it would span the level from that line onwards, defined inside another block it should only handle exceptions that are thrown inside the specific block:

    void Method() {
      catch with => ErrorHandler;
      if (someCondition) {
        catch with => AnotherErrorHandler;
        SomeMethodThrowingAnException();
      }
      SomeMethodThrowingAnException();
    }

Should therefore be lowered to:

    void Method() {
      try {
        if (someCondition) {
          try {
            SomeMethodThrowingAnException();
          } catch (Exception ex2) {
            // do something else
          }
        }
        SomeMethodThrowingAnException();
      } catch (Exception ex) {
        // do something with the exception
      }
    }

Combined with using statements it can look like this:

    catch with => (e) => Console.WriteLine(e.Message);
    using var file = System.IO.File.Open("somefile.txt", System.IO.FileMode.CreateNew);

In the lowered version it may look like this:

    try {
      var file = File.Open("somefile.txt", FileMode.CreateNew);
      try {
      } finally {
        if (fileStream != null) {
          ((IDisposable)fileStream).Dispose();
        }
      }
    } catch (Exception ex) {
      Console.WriteLine(ex.Message);
    }

It might be optimized by moving the declaration out of the try-block and adding a finally block as long as the lifetime of the FileStream in th is example is not extended. Nevertheless I do not see a lot of value in this optimization.

# Potential additions

## Specific exception types

The proposed syntax allows for type checking inside the handler like this:

    void CheckException( Exception ex ) {
      switch ( ex ) {
        case InvalidOperationException:
          Console.WriteLine("invalid operation");
          break;
        case not null:
          Console.WriteLine("generic exception");
          break;
      }
    }

Therefore adding logic for different exception types is already possible. It would also be possible to adjust the proposed syntax to be able to provide different handlers per type like this:

    catch with => (e) => Console.WriteLine("generic exception");
    catch<InvalidOperationException> with => (e) => Console.WriteLine("invalid operation");
    SomeMethodThrowingAnException();

The "generic" handler definition would rethrow all not explicitly handled exceptions in a default handler:

    try {
      try {
        SomeMethodThrowingAnException();
      } catch (InvalidOperationException iex) {
        Console.WriteLine("invalid operation");
      } catch {
        throw;
      }
    } catch (Exception ex) {
      Console.WriteLine("generic exception");
    }
    
Note: due to scoping, the definition of the handlers should go from generic to specific.

# Potential shortcomings

## Finally

In the simple approach presented above support for finally is available. For a first iteration I would keep the changes simple and actually recommend using the traditional syntax for use cases with the requirement of a manually implemented finally-block.

A possible (adding slightly more complexity to the proposal) change to this syntax is allowing for a second overload in the form of a tuple having an action for the finally block:

    catch with => ( (e) => SomeAction(e), () => SomeOtherAction() )
    SomeMethodThrowingAnException();

With lowering to:

    try {
      SomeMethodThrowingAnException();
    } catch (Exception ex) {
      SomeAction(e);
    } finally {
      SomeOtherAction();
    }

**Comment by HaloFour** [comment]
I don't particularly care for inverting the order of declaration and I really don't like having the `try` scope be implicit.  It should be uncommon to wrap the entire body of a method in a `try` scope, but this proposal encourages doing that.  Despite all of that, it doesn't seem to be appreciably more concise than current `try`/`catch` blocks, you've mostly just adopted different whitespace formatting.

**Comment by sascha-andres** [comment]
> I really don't like having the try scope be implicit.

I don't think it is implicit in that there is a clear rule: starting from the current line till the end of the current scope

> It should be uncommon to wrap the entire body of a method in a try scope

Totally agree that it should be uncommon to wrap the  entire body of of method (though I've seen that way too often unfortunately and I personally think this is a code smell). But you can limit the scope of the error handler by starting on a later line or opening it in a deeper scope

**Reply by theunrepentantgeek** [reply]
You're trading explicitness, where the scope is naturally apparent from the structure of the code, for implicitness, where readers have to carefully read and think about the code in order to understand it. 

Surely it's hard enough to read and reason about unfamiliar code already, we shouldn't be going out of our way to make things more difficult.

> To improve readability common approaches are to limit nesting and to reduce line of code required to code a feature. 

I don't think limiting nesting _ever_ been an explicit goal of the C# language design team, though it has been a welcome side effect of some of the work that has been done.

The closest might have been the introduction of expression-bodied-members (`=>` for properties and methods) but the goal there was to streamline the extremely common case where those bodies had a single expression/statement. By contrast, **try**..**catch** statements are relatively rare.

**Reply by CyrusNajmabadi** [reply]
> I don't think limiting nesting ever been an explicit goal of the C# language design team, though it has been a welcome side effect of some of the work that has been done.

We want common stuff to be pleasant to write.  Uncommon stuff (i.e. "exceptional" stuff) is good to stand out IMO :)
